src/library/dao/cratedao.cpp

```diff
@@ -18,6 +18,32 @@ CrateDAO::~CrateDAO() {
 
 void CrateDAO::initialize() {
     qDebug() << "CrateDAO::initialize()";
+
+    //get the count to allocate HashMap
+    int tracksInCratesCount = 0;
+    QSqlQuery query(m_database);
+    query.prepare("SELECT COUNT(*) from " CRATE_TRACKS_TABLE);
+
+    if (!query.exec()) {
+        LOG_FAILED_QUERY(query);
+    }
+
+    tracksInCratesCount = query.value(0).toInt();
+
+    m_cratesTrackIsIn.reserve(tracksInCratesCount);
+
+    //now fetch all Tracks from all crates and insert them into the hashmap
+    query.prepare("SELECT track_id, crate_id from " CRATE_TRACKS_TABLE);
+    if (!query.exec()) {
+        LOG_FAILED_QUERY(query);
+    }
+
+    const int trackIdColumn = query.record().indexOf("track_id");
+    const int crateIdColumn = query.record().indexOf("crate_id");
+    while (query.next()) {
+        m_cratesTrackIsIn.insert(query.value(crateIdColumn).toInt(),
+                                 query.value(trackIdColumn).toInt());
+    }
```

Comment:
This is finally a decision of taste. If you have a working solution that has a reasonable CPU load it is a good one. 
It is no issue to change the way of implementation later if we actually found performance issues. 

This and to quoted bugs can be solved with an in memory table and with an sql qerry. 

Outlook for the future: 
Nazars branch provides a non blocking access to sql database and will be part of Mixxx 1.13. The result is prepared in a separate thread and passed to the GUI thread in a new slot. 
Once this is integrated we have the option to block the GUI until the result is calculated from an in memory table or delay the GUI update in a non blocking way by until the SQL query is finished. 
Such a decoration feature like this is IMHO a candidate for the later option. 

Without the non blocking SQL access, an internal table access will block the GUI shorter, than an sql query, but we need more CPU to keep the table up to date and there is an amount of additional code maintenance and complexity. For now it looks like the internal table wins. 
Do you have a working solution for both? Take the on that suits more to you.   
I have no Idea which performance tweaks are worth the works. 

I think I will test it on my atom net-book, once we consider this as merge-able.     
  
By the way: Nazars branch needs a new maintainer. The latest version is already working, but it falls behind the master branch. The next steps to do is to merge it with master by crawling up the master branch one by one database related commit. You will find it here: https://github.com/mixxxdj/mixxx/pull/73